src/dune_rules/merlin.ml

```diff
@@ -61,36 +56,40 @@ module Processed = struct
           [ make_directive "FLG" (Sexp.List (List.map ~f:to_atom flags)) ]
       in
       match pp with
+      | None -> flags
       | Some (pp_flag, pp_args) ->
         make_directive "FLG" (Sexp.List [ Atom pp_flag; Atom pp_args ]) :: flags
-      | None -> flags
     in
     let suffixes =
-      Extensions.Set.to_list extensions
-      |> List.map ~f:(fun (impl, intf) ->
-             make_directive "SUFFIX"
-               (Sexp.Atom (Printf.sprintf "%s %s" impl intf)))
+      List.map extensions ~f:(fun { Ml_kind.Dict.impl; intf } ->
+          make_directive "SUFFIX" (Sexp.Atom (Printf.sprintf "%s %s" impl intf)))
     in
     Sexp.List
       (List.concat [ exclude_query_dir; obj_dirs; src_dirs; flags; suffixes ])
 
-  let rec get config ~filename =
-    let file = Filename.remove_extension filename in
-    match Option.map (String.Map.find config filename) ~f:to_sexp with
-    | Some result -> Some result
-    | None when String.equal file filename -> None
-    | None -> get config ~filename:file
+  let get { modules; config } ~filename =
+    let fname = Filename.remove_extension filename in
+    if
+      List.exists modules ~f:(fun m ->
+          let fname' = Module_name.to_string m |> String.lowercase in
+          String.equal fname fname')
+    then
+      Some (to_sexp config)
+    else
+      None
 
   let print_file path =
     match load_file path with
     | None -> Printf.eprintf "No merlin config found"
-    | Some t ->
+    | Some { modules; config } ->
       let pp =
-        String.Map.to_list t
-        |> Pp.concat_map ~f:(fun (name, config) ->
-               let sexp = to_sexp config in
-               let open Pp.O in
-               Pp.vbox (Pp.text name ++ Pp.cut ++ Sexp.pp sexp) ++ Pp.newline)
+        let sexp = to_sexp config in
+        let open Pp.O in
+        Pp.vbox (Sexp.pp sexp)
+        ++ Pp.newline
+        ++ Pp.vbox
+             (Pp.concat_map modules ~f:(fun m ->
+                  Pp.text (Module_name.to_string m)))
```

Comment:
This is use for debugging and testing only, and it feels less useful to have a module's name printed after its configuration (and it brings lot of small changes to the tests). Is there a specific reason why you chose to switch them ?

Reply:
I just thought it made the tests neater. Now we aren't printing the same configuration over and over.

Reply:
I agree that merging identical configuration is better, I was just complaining about the module names being printed after and not before the configuration.

However I had to re-duplicate for `Per_module` pp configs #4092.